### PR TITLE
KAFKA-5096: Log invalid user configs and use defaults

### PR DESCRIPTION
**Pull request by johnma14**
Kafka Streams does not allow users to modify some consumer configurations.
Currently, it does not allow modifying the value of 'enable.auto.commit'.
If the user modifies this property, currently an exception is thrown.
The following changes were made in this patch:
- Defined a new array 'NON_CONFIGURABLE_CONSUMER_CONFIGS' to hold the names
  of the configuration parameters that is not allowed to be modified
- Defined a new method 'checkIfUnexpectedUserSpecifiedConsumerConfig' to
  check if user overwrote the values of any of the non configurable configuration
  parameters. If so, then log a warning message and reset the default values
- Updated the javadoc to include the configuration parameters that cannot be
  modified by users.
- Updated the corresponding tests in StreamsConfigTest.java to reflect the changes
  made in StreamsConfig.java

### CODE REVIEW

**Comment by johnma14** [comment]
Thiis patch is a continuation of https://github.com/apache/kafka/pull/2938/commits/435cd3d2efdc0f70cc361f24648ce5cb4cd3c680 after making the recommended changes. Unfortunately, I had to close the previous one since my VM crashed and it messed up my env.

**Comment by johnma14** [comment]
retest this please

**Comment by johnma14** [comment]
retest this please

**Comment by johnma14** [comment]
retest this please

**Comment by johnma14** [comment]
@mjsax  This patch is ready for review whenever it is a good time for you. Appreciate your time and feedback. This has the changes requested in the following review: https://github.com/apache/kafka/pull/2938/files. I had to close that one because I had some trouble uploading the latest changes to that branch after my VM crashed. Thank you.

**Comment by mjsax** [comment]
Thank @johnma14 . Btw: https://issues.apache.org/jira/browse/KAFKA-4923 got merged yesterday, thus, the scope of this JIRA is enlarged by a few more configs we don't allow to overwrite if EOS is enabled. Can you please rebase to `trunk` and update this PR accordingly? Thanks a lot.

**Comment by johnma14** [comment]
@mjsax I just noticed that. Will make the necessary changes and upload another patch. Thank you.

**Comment by johnma14** [comment]
@mjsax This patchset is ready for review whenever it is a good time for you. It includes the latest changes made to the StreamsConfig.java class related to EOS. Thank you.

**Review by mjsax — APPROVED** [review]
LGTM. Thanks for the patch @johnma14 (and sorry for the wait -- got side tracked)

@bill @enothereska @dguy Call for second review.

@guozhangwang Call for merging for `0.11` (if possible -- it's tight and not a must -- but might be nice to get into the release)

**Comment by johnma14** [comment]
Thanks @mjsax. Appreciate it

**Review by enothereska — APPROVED** [review]

**Comment by johnma14** [comment]
@guozhangwang Wondering is you have some time to take a look at this patch. It has been approved by @mjsax and @enothereska. Thank you and appreciate your time.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Refer to the `processing.guarantee` config here.

**Review by guozhangwang — COMMENTED** [review]
@johnma14 Thanks for the patch and sorry for getting late on it! I really appreciate the thorough description on the PR. Just one minor comment otherwise LGTM.

**Comment by johnma14** [comment]
@guozhangwang Thank you for reviewing the changes. I have pushed another patch set reflecting the changes your recommended. Appreciate your feedback and time. Thank you.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5469/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5484/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by mjsax** [comment]
Retest this please.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5490/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5475/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by johnma14** [comment]
@guozhangwang, @mjsax @enothereska Wondering if you have some time to take a look at this patch. Made the changes recommended by @guozhangwang. Thank you and appreciate your time.

**Review by mjsax — APPROVED** [review]
LGTM

**Comment by johnma14** [comment]
@guozhangwang @enothereska  Wondering if you have some time to take a look at this patch. Thank you and appreciate your time.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6342/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by johnma14** [comment]
@mjsax thank you for your review.

**Comment by johnma14** [comment]
retest this please

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6358/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6343/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by johnma14** [comment]
@mjsax @guozhangwang @enothereska Just putting in another request for review. Thanks.

**Comment by dguy** [comment]
retest this please

**Review by dguy — APPROVED** [review]
Thanks @johnma14, LGTM. Sorry this has been sitting here for a while. I'll merge it once the tests pass again.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6477/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6492/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by johnma14** [comment]
Thank you @dguy. Appreciate it.

**Comment by johnma14** [comment]
retest this please

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6494/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6479/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by dguy** [comment]
Thanks @johnma14 merged to trunk

**Pull request closed without merging**
